### PR TITLE
Issue #SB-3374 feat: Sunbird-OpenSaber-Integration

### PR DESCRIPTION
**Pull request by JaikumarSRajan**
Added - UserExtension.java
Modified - User.java, JsonKey.java, ResponseCode.java, ResponseMessage.java, externalresource.properties, cassandratablecolumn.properties

jackson-databind version update to 2.9.4 in
actor-core/pom.xml
actor-util/pom.xml
cassandra-dac/pom.xml
common-util/pom.xml
models/pom.xml
request-validator/pom.xml

### CODE REVIEW

File: models/src/main/java/org/sunbird/models/user/User.java [hunk 1]
```diff
@@ -55,6 +55,24 @@
   private String externalId;
   private String channel;
   private String loginId;
+  private String registryId;
+  private Map<String, Object> registry;
```
Comment (bvinayakumar) [inline models/src/main/java/org/sunbird/models/user/User.java]: Don't add registry (fields) within Sunbird User model.

File: extensions/src/main/java/org/sunbird/extension/user/UserExtension.java [hunk 2]
```diff
@@ -3,16 +3,38 @@
...
+   * @param userProfileMap User profile information
+   */
+  public void create(Map<String, Object> userProfileMap);
+
+  /**
+   * Reads user profile information.
+   *
+   * @param userProfileMap User profile information
+   */
+  public void read(Map<String, Object> userProfileMap);
```
Comment (bvinayakumar) [inline extensions/src/main/java/org/sunbird/extension/user/UserExtension.java]: read API should accept userIdMap (User ID information) and return userProfileMap (user profile information).

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 3]
```diff
@@ -317,6 +317,27 @@
...
         "Cassandra connection establishment failed in {0} mode.";
     String COMMON_ATTRIBUTE_MISMATCH = "{0} mismatch of {1} and {2}";
+    String JSON_TRANSFORM_INVALID_TYPE_CONFIG =
+        "Invalid type configuration for JSON transformation";
+    String JSON_TRANSFORM_INVALID_DATE_FORMAT =
+        "Invalid date format configuration for JSON transformation";
+    String JSON_TRANSFORM_FAILED_INVALID_INPUT =
+        "JSON Transformation failed due to invalid input data format";
+    String JSON_TRANSFORM_ENUM_VALUES_MISSING = "Enum values missing in the configuration";
+    String JSON_TRANSFORM_BASIC_CONFIG_MISSING =
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Mandatory configuration (toFieldName, fromType or toType) is missing for field {0}.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 4]
```diff
@@ -317,6 +317,27 @@
     String CASSANDRA_CONNECTION_ESTABLISHMENT_FAILED =
         "Cassandra connection establishment failed in {0} mode.";
     String COMMON_ATTRIBUTE_MISMATCH = "{0} mismatch of {1} and {2}";
+    String JSON_TRANSFORM_INVALID_TYPE_CONFIG =
+        "Invalid type configuration for JSON transformation";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: JSON transformation failed as invalid type configuration found for field {0}.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 5]
```diff
@@ -317,6 +317,27 @@
...
+    String JSON_TRANSFORM_INVALID_DATE_FORMAT =
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: JSON transformation failed as invalid date format configuration found for field {0}.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 6]
```diff
@@ -317,6 +317,27 @@
...
+        "Invalid date format configuration for JSON transformation";
+    String JSON_TRANSFORM_FAILED_INVALID_INPUT =
+        "JSON Transformation failed due to invalid input data format";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: ERROR_JSON_TRANSFORM_INVALID_INPUT
JSON transformation failed as invalid input provided for field {0}.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 7]
```diff
@@ -317,6 +317,27 @@
...
+    String JSON_TRANSFORM_ENUM_VALUES_MISSING = "Enum values missing in the configuration";
+    String JSON_TRANSFORM_BASIC_CONFIG_MISSING =
+        "one or more basic config fields(toFieldName, fromType, toType) are missing ";
+    String USER_REGISTRY_LOAD_CONFIG_FAILED = "Loading of User Registry Configurations failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: ERROR_USER_REGISTRY_LOAD_CONFIG:
Loading failed for user registry configuration file {0}.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 8]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_CLIENT_CREATION_FAILED = "User Registry Client Creation Failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: User registry client creation failed.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 9]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_ADD_ENTITY_FAILED = "User Registry Add Entity Failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: User registry add entity API failed.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 10]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_READ_ENTITY_FAILED = "User Registry Read Entity Failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: User registry read entity API failed.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 11]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_UPDATE_ENTITY_FAILED = "User Registry Update Entity Failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: User registry update entity API failed.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 12]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_DELETE_ENTITY_FAILED = "User Registry Delete Entity Failed";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: User registry delete entity API failed.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 13]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_PARSE_RESPONSE_FAILED = "Error while parsing response from registry";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Error while parsing response from registry.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 14]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_USER_TYPE_BLANK = "User Registry - UserType is blank";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: ERROR_USER_REGISTRY_USER_TYPE_BLANK:
Request failed as user type is blank.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 15]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_MAIN_PROVIDER_NOT_CONFIGURED =
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: ERROR_USER_REGISTRY_MAIN_PROVIDER_NOT_CONFIGURED:
Request failed as provider not configured for user ID.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 16]
```diff
@@ -317,6 +317,27 @@
...
+        "User Registry - Main Provider is not configured";
+    String USER_REGISTRY_UNIQUE_ID_BLANK = "User Registry - RegistryId is blank";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Request failed as registryId is not provided.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 17]
```diff
@@ -317,6 +317,27 @@
...
+    String USER_REGISTRY_ACCESS_TOKEN_BLANK = "User Registry - User Access Token is blank";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Don't throw this exception as it is not mandatory for user create.

File: extensions/src/main/java/org/sunbird/extension/user/UserExtension.java [hunk 18]
```diff
@@ -3,16 +3,39 @@
...
+   *
+   * @param userProfileMap User profile information
+   */
+  public void create(Map<String, Object> userProfileMap);
+
+  /**
+   * Reads user profile information.
+   *
+   * @param userIdMap UserId Information
+   * @return userProfileMap Map<String, Object>
```
[the remark this answers is not in the document]
Comment (JaikumarSRajan) [reply extensions/src/main/java/org/sunbird/extension/user/UserExtension.java]: Review Comment Incorporated

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java (hunk 16 above)
Comment (JaikumarSRajan) [reply common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Review comment incorporated

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java (hunk 15 above)
Comment (JaikumarSRajan) [reply common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Review comment incorporated

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java (hunk 14 above)
Comment (JaikumarSRajan) [reply common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Review comment incorporated

**Pull request merged**